pkg/operator/workloads/logs.go

```diff
@@ -133,52 +143,164 @@ func ReadLogs(appName string, workloadID string, verbose bool, socket *websocket
 	}
 }
 
-func getKubectlLogs(pod *kcore.Pod, verbose bool, wrotePending bool, previous bool, socket *websocket.Conn) {
-	cmdPath := "/usr/local/bin/kubectl"
-
-	if k8s.GetPodStatus(pod) == k8s.PodStatusPending {
-		if !wrotePending {
-			if !writeSocket("\nPending", socket) {
-				return
-			}
+func getKubectlLogs(pods []kcore.Pod, verbose bool, wrotePending bool, previous bool, socket *websocket.Conn) {
+	isAllPending := true
+	for _, pod := range pods {
+		if k8s.GetPodStatus(&pod) != k8s.PodStatusPending {
+			isAllPending = false
+			break
 		}
-		config.Kubernetes.WaitForPodRunning(pod.Name, 1)
 	}
 
-	args := []string{"kubectl", "-n=" + config.Cortex.Namespace, "logs", "--follow=true"}
-	if previous {
-		args = append(args, "--previous")
-	}
-
-	args = append(args, pod.Name)
-	if pod.Labels["workloadType"] == WorkloadTypeAPI && pod.Labels["userFacing"] == "true" {
-		args = append(args, apiContainerName)
+	if isAllPending {
+		if !writeSocket("\nPending", socket) {
+			return
+		}
 	}
 
 	outr, outw, err := os.Pipe()
 	if err != nil {
 		errors.Panic(err, "logs", "kubectl", "os.pipe")
 	}
 	defer outr.Close()
-	defer outw.Close()
 
 	inr, inw, err := os.Pipe()
 	if err != nil {
 		errors.Panic(err, "logs", "kubectl", "os.pipe")
 	}
-	defer inr.Close()
 	defer inw.Close()
 
-	process, err := os.StartProcess(cmdPath, args, &os.ProcAttr{
-		Files: []*os.File{inr, outw, outw},
-	})
+	podCheckCancel := make(chan struct{})
+	defer close(podCheckCancel)
+	socketWriterError := make(chan error, 1)
+	go podCheck(podCheckCancel, socketWriterError, pods, previous, inr, outw)
+	go pumpStdout(socket, socketWriterError, outr, verbose, true)
+	pumpStdin(socket, inw)
+	podCheckCancel <- struct{}{}
+}
+
+func startKubectlProcess(pod kcore.Pod, previous bool, attrs *os.ProcAttr) (*os.Process, error) {
+	cmdPath := "/bin/bash"
+
+	kubectlArgs := []string{"kubectl", "-n=" + config.Cortex.Namespace, "logs", "--follow=true"}
+	if previous {
+		kubectlArgs = append(kubectlArgs, "--previous")
+	}
+
+	identifier := pod.Name
+	kubectlArgs = append(kubectlArgs, pod.Name)
+	if pod.Labels["workloadType"] == WorkloadTypeAPI && pod.Labels["userFacing"] == "true" {
+		kubectlArgs = append(kubectlArgs, apiContainerName)
+		kubectlArgs = append(kubectlArgs, "--since="+podCheckInterval.String())
```

Comment:
I preferred `--since` to handle the `podCheckInterval` , ie, if podcheck goes to sleep and a new pod is added right away, we should replay the last `podCheckInterval` seconds in logs. We can talk about which makes more sense.